File: bindings/ruby/lib/proven_servers/tls.rb
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
#
# Ruby bindings for the proven-tls Zig FFI.
#
# Wraps the C-ABI functions for TLS handshake lifecycle,
# cipher negotiation, certificate validation, and data transfer.

# frozen_string_literal: true

require "ffi"

module ProvenServers
  # TLS protocol bindings matching the Idris2 ABI.
  #
  # @example
  #   ProvenServers::Tls.with_context do |ctx|
  #     ctx.client_hello
  #     ctx.server_hello
  #     ctx.negotiate(ProvenServers::Tls::CipherSuite::AES_256_GCM_SHA384)
  #     ctx.validate_cert(ProvenServers::Tls::CertStatus::VALID)
  #     ctx.complete_handshake
  #     ctx.send_data(1024)
  #     ctx.shutdown
  #   end
  module Tls
    extend FFI::Library

    FFILoader.load_protocol_library(self, "tls")

    # TLS handshake lifecycle states matching Idris2 ABI tags.
    module TlsState
      IDLE               = 0
      CLIENT_HELLO       = 1
      SERVER_HELLO       = 2
      NEGOTIATED         = 3
      HANDSHAKE_COMPLETE = 4
      APPLICATION_DATA   = 5
      SHUTDOWN           = 6
      CLOSED             = 7
    end

    # TLS protocol versions matching Idris2 ABI tags.
    module TlsVersion
      TLS_1_2 = 0
      TLS_1_3 = 1
    end

    # TLS cipher suites matching Idris2 ABI tags.
    module CipherSuite
      AES_128_GCM_SHA256       = 0
      AES_256_GCM_SHA384       = 1
      CHACHA20_POLY1305_SHA256 = 2
      AES_128_CCM_SHA256       = 3
    end

    # Certificate validation status matching Idris2 ABI tags.
    module CertStatus
      UNCHECKED         = 0
      VALID             = 1
      EXPIRED           = 2
      REVOKED           = 3
      SELF_SIGNED       = 4
      UNKNOWN_CA        = 5
      HOSTNAME_MISMATCH = 6
    end

    # TLS alert levels matching Idris2 ABI tags.
    module AlertLevel
      WARNING = 0
      FATAL   = 1
    end

    # FFI function declarations.
    attach_function :tls_create,             [:uint8, :uint8], :int
    attach_function :tls_destroy,            [:int], :void
    attach_function :tls_state,              [:int], :uint8
    attach_function :tls_version,            [:int], :uint8
    attach_function :tls_cipher_suite,       [:int], :uint8
    attach_function :tls_cert_status,        [:int], :uint8
    attach_function :tls_is_resumed,         [:int], :uint8
    attach_function :tls_bytes_sent,         [:int], :uint64
    attach_function :tls_bytes_received,     [:int], :uint64
    attach_function :tls_client_hello,       [:int], :uint8
    attach_function :tls_server_hello,       [:int], :uint8
    attach_function :tls_negotiate,          [:int, :uint8], :uint8
    attach_function :tls_validate_cert,      [:int, :uint8], :uint8
    attach_function :tls_complete_handshake, [:int], :uint8
    attach_function :tls_send_data,          [:int, :uint64], :uint8
    attach_function :tls_receive_data,       [:int, :uint64], :uint8
    attach_function :tls_rekey,              [:int], :uint8
    attach_function :tls_shutdown,           [:int], :uint8
    attach_function :tls_send_alert,         [:int, :uint8], :uint8
    attach_function :tls_abi_version,        [], :uint32
    attach_function :tls_can_transition,     [:uint8, :uint8], :uint8

    # TLS session context wrapping a Zig FFI slot.
    class Context
      attr_reader :slot

      def initialize(slot)
        @slot = slot
        @destroyed = false
      end

      # @param version [Integer] TlsVersion tag (default: TLS_1_3)
      # @param cipher_suite [Integer] CipherSuite tag (default: AES_256_GCM_SHA384)
      # @return [Context]
      def self.create(version: TlsVersion::TLS_1_3, cipher_suite: CipherSuite::AES_256_GCM_SHA384)
        slot = ProvenServers.check_slot(Tls.tls_create(version, cipher_suite))
        new(slot)
      end

      # @return [void]
      def destroy
        return if @destroyed

        Tls.tls_destroy(@slot)
        @destroyed = true
      end

      # @return [Integer, nil] TlsState tag
      def state
        tag = Tls.tls_state(@slot)
        tag <= 7 ? tag : nil
      end

      # @return [Integer, nil] TlsVersion tag
      def version
        tag = Tls.tls_version(@slot)
        tag <= 1 ? tag : nil
      end

      # @return [Integer, nil] CipherSuite tag
      def cipher_suite
        tag = Tls.tls_cipher_suite(@slot)
        tag <= 3 ? tag : nil
      end

      # @return [Integer, nil] CertStatus tag
      def cert_status
        tag = Tls.tls_cert_status(@slot)
        tag <= 6 ? tag : nil
      end

      # @return [Boolean]
      def resumed?        = Tls.tls_is_resumed(@slot) == 1
      # @return [Integer]
      def bytes_sent      = Tls.tls_bytes_sent(@slot)
      # @return [Integer]
      def bytes_received  = Tls.tls_bytes_received(@slot)

      # @return [void]
      def client_hello      = ProvenServers.check_status(Tls.tls_client_hello(@slot))
      # @return [void]
      def server_hello      = ProvenServers.check_status(Tls.tls_server_hello(@slot))

      # Negotiate a cipher suite.
      #
      # @param cipher_suite [Integer] CipherSuite tag
      # @return [void]
      def negotiate(cipher_suite)
        ProvenServers.check_status(Tls.tls_negotiate(@slot, cipher_suite))
      end

      # Validate the peer certificate.
      #
      # @param status [Integer] CertStatus tag
      # @return [void]
      def validate_cert(status)
        ProvenServers.check_status(Tls.tls_validate_cert(@slot, status))
      end

      # @return [void]
      def complete_handshake = ProvenServers.check_status(Tls.tls_complete_handshake(@slot))

      # Record data sent.
      #
      # @param length [Integer]
      # @return [void]
      def send_data(length)
        ProvenServers.check_status(Tls.tls_send_data(@slot, length))
      end

      # Record data received.
      #
      # @param length [Integer]
      # @return [void]
      def receive_data(length)
        ProvenServers.check_status(Tls.tls_receive_data(@slot, length))
      end

      # @return [void]
      def rekey     = ProvenServers.check_status(Tls.tls_rekey(@slot))
      # @return [void]
      def shutdown  = ProvenServers.check_status(Tls.tls_shutdown(@slot))

      # Send a TLS alert.
      #
      # @param level [Integer] AlertLevel tag
      # @return [void]
      def send_alert(level)
        ProvenServers.check_status(Tls.tls_send_alert(@slot, level))
      end
    end

    # @yield [Context]
    # @return [Object]
    def self.with_context(version: TlsVersion::TLS_1_3, cipher_suite: CipherSuite::AES_256_GCM_SHA384)
      ctx = Context.create(version: version, cipher_suite: cipher_suite)
      begin
        yield ctx
      ensure
        ctx.destroy
      end
    end

    # @return [Integer]
    def self.abi_version = tls_abi_version
    # @param from [Integer] @param to [Integer] @return [Boolean]
    def self.can_transition?(from, to) = tls_can_transition(from, to) == 1
  end
end

File: bindings/elixir/lib/proven_servers/tls.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
defmodule ProvenServers.Tls do
  @moduledoc """
  TLS protocol types for the proven-servers ABI.

  Models TLS 1.2 and 1.3 protocol types based on RFC 8446 (TLS 1.3)
  and RFC 5246 (TLS 1.2). Tag values are provisional pending Idris2
  ABI formalisation.

  ## Handshake State Machine (TLS 1.3)

      Start -> WaitServerHello -> WaitEncryptedExtensions ->
      WaitCertRequest -> WaitCert -> WaitFinished -> Connected
  """

  # ===========================================================================
  # TLS Constants
  # ===========================================================================

  @doc "Standard HTTPS port (443)."
  @spec https_port() :: non_neg_integer()
  def https_port, do: 443

  @doc "Maximum TLS record size in bytes (16384, RFC 8446 Section 5.1)."
  @spec max_record_size() :: non_neg_integer()
  def max_record_size, do: 16_384

  @doc "Maximum TLS record size with padding (16640, RFC 8446 Section 5.4)."
  @spec max_record_size_with_padding() :: non_neg_integer()
  def max_record_size_with_padding, do: 16_640

  # ===========================================================================
  # TlsVersion (provisional tags 0-3)
  # ===========================================================================

  @typedoc "TLS protocol versions."
  @type tls_version :: :tls10 | :tls11 | :tls12 | :tls13

  @version_tags %{tls10: 0, tls11: 1, tls12: 2, tls13: 3}
  @tag_to_version Map.new(@version_tags, fn {k, v} -> {v, k} end)

  @version_wires %{tls10: 0x0301, tls11: 0x0302, tls12: 0x0303, tls13: 0x0304}
  @version_strings %{tls10: "TLS 1.0", tls11: "TLS 1.1", tls12: "TLS 1.2", tls13: "TLS 1.3"}

  @doc "Decode from a C-ABI tag value."
  @spec version_from_tag(non_neg_integer()) :: {:ok, tls_version()} | :error
  def version_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 3 do
    {:ok, Map.fetch!(@tag_to_version, tag)}
  end
  def version_from_tag(_tag), do: :error

  @doc "Encode to the C-ABI tag value."
  @spec version_to_tag(tls_version()) :: non_neg_integer()
  def version_to_tag(v) when is_map_key(@version_tags, v), do: Map.fetch!(@version_tags, v)

  @doc "TLS wire protocol version bytes (major.minor as u16)."
  @spec version_wire_value(tls_version()) :: non_neg_integer()
  def version_wire_value(v) when is_map_key(@version_wires, v), do: Map.fetch!(@version_wires, v)

  @doc """
  Human-readable version string.

  ## Examples

      iex> ProvenServers.Tls.version_to_string(:tls13)
      "TLS 1.3"
  """
  @spec version_to_string(tls_version()) :: String.t()
  def version_to_string(v) when is_map_key(@version_strings, v), do: Map.fetch!(@version_strings, v)

  @doc "Whether this version is considered secure (TLS 1.2+ per RFC 8996)."
  @spec version_secure?(tls_version()) :: boolean()
  def version_secure?(v) when v in [:tls12, :tls13], do: true
  def version_secure?(_v), do: false

  # ===========================================================================
  # HandshakeType (RFC 8446 Section 4, provisional tags 0-10)
  # ===========================================================================

  @typedoc "TLS handshake message types (RFC 8446 Section 4)."
  @type handshake_type ::
          :client_hello | :server_hello | :new_session_ticket | :end_of_early_data
          | :encrypted_extensions | :certificate | :certificate_request
          | :certificate_verify | :finished | :key_update | :message_hash

  @handshake_type_tags %{
    client_hello: 0, server_hello: 1, new_session_ticket: 2,
    end_of_early_data: 3, encrypted_extensions: 4, certificate: 5,
    certificate_request: 6, certificate_verify: 7, finished: 8,
    key_update: 9, message_hash: 10
  }
  @tag_to_handshake_type Map.new(@handshake_type_tags, fn {k, v} -> {v, k} end)

  @handshake_type_wires %{
    client_hello: 1, server_hello: 2, new_session_ticket: 4,
    end_of_early_data: 5, encrypted_extensions: 8, certificate: 11,
    certificate_request: 13, certificate_verify: 15, finished: 20,
    key_update: 24, message_hash: 254
  }

  @doc "Decode from a C-ABI tag value."
  @spec handshake_type_from_tag(non_neg_integer()) :: {:ok, handshake_type()} | :error
  def handshake_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 10 do
    {:ok, Map.fetch!(@tag_to_handshake_type, tag)}
  end
  def handshake_type_from_tag(_tag), do: :error

  @doc "Encode to the C-ABI tag value."
  @spec handshake_type_to_tag(handshake_type()) :: non_neg_integer()
  def handshake_type_to_tag(ht) when is_map_key(@handshake_type_tags, ht) do
    Map.fetch!(@handshake_type_tags, ht)
  end

  @doc "RFC wire value for the handshake type."
  @spec handshake_type_wire_value(handshake_type()) :: non_neg_integer()
  def handshake_type_wire_value(ht) when is_map_key(@handshake_type_wires, ht) do
    Map.fetch!(@handshake_type_wires, ht)
  end

  # ===========================================================================
  # CipherSuite (TLS 1.3, provisional tags 0-4)
  # ===========================================================================

  @typedoc "TLS 1.3 cipher suites (RFC 8446 Section 9.1)."
  @type cipher_suite ::
          :aes_128_gcm_sha256 | :aes_256_gcm_sha384 | :chacha20_poly1305_sha256
          | :aes_128_ccm_sha256 | :aes_128_ccm_8_sha256

  @cipher_suite_tags %{
    aes_128_gcm_sha256: 0, aes_256_gcm_sha384: 1,
    chacha20_poly1305_sha256: 2, aes_128_ccm_sha256: 3, aes_128_ccm_8_sha256: 4
  }
  @tag_to_cipher_suite Map.new(@cipher_suite_tags, fn {k, v} -> {v, k} end)

  @cipher_suite_ianas %{
    aes_128_gcm_sha256: 0x1301, aes_256_gcm_sha384: 0x1302,
    chacha20_poly1305_sha256: 0x1303, aes_128_ccm_sha256: 0x1304, aes_128_ccm_8_sha256: 0x1305
  }

  @cipher_suite_names %{
    aes_128_gcm_sha256: "TLS_AES_128_GCM_SHA256",
    aes_256_gcm_sha384: "TLS_AES_256_GCM_SHA384",
    chacha20_poly1305_sha256: "TLS_CHACHA20_POLY1305_SHA256",
    aes_128_ccm_sha256: "TLS_AES_128_CCM_SHA256",
    aes_128_ccm_8_sha256: "TLS_AES_128_CCM_8_SHA256"
  }

  @doc "Decode from a C-ABI tag value."
  @spec cipher_suite_from_tag(non_neg_integer()) :: {:ok, cipher_suite()} | :error
  def cipher_suite_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_cipher_suite, tag)}
  end
  def cipher_suite_from_tag(_tag), do: :error

  @doc "Encode to the C-ABI tag value."
  @spec cipher_suite_to_tag(cipher_suite()) :: non_neg_integer()
  def cipher_suite_to_tag(cs) when is_map_key(@cipher_suite_tags, cs) do
    Map.fetch!(@cipher_suite_tags, cs)
  end

  @doc "IANA cipher suite value (RFC 8446 Appendix B.4)."
  @spec cipher_suite_iana_value(cipher_suite()) :: non_neg_integer()
  def cipher_suite_iana_value(cs) when is_map_key(@cipher_suite_ianas, cs) do
    Map.fetch!(@cipher_suite_ianas, cs)
  end

  @doc "OpenSSL-style cipher suite name."
  @spec cipher_suite_name(cipher_suite()) :: String.t()
  def cipher_suite_name(cs) when is_map_key(@cipher_suite_names, cs) do
    Map.fetch!(@cipher_suite_names, cs)
  end

  # ===========================================================================
  # AlertLevel (RFC 8446 Section 6, provisional tags 0-1)
  # ===========================================================================

  @typedoc "TLS alert levels (RFC 8446 Section 6.1)."
  @type alert_level :: :warning | :fatal

  @doc "Decode from a C-ABI tag value."
  @spec alert_level_from_tag(non_neg_integer()) :: {:ok, alert_level()} | :error
  def alert_level_from_tag(0), do: {:ok, :warning}
  def alert_level_from_tag(1), do: {:ok, :fatal}
  def alert_level_from_tag(_tag), do: :error

  @doc "Encode to the C-ABI tag value."
  @spec alert_level_to_tag(alert_level()) :: non_neg_integer()
  def alert_level_to_tag(:warning), do: 0
  def alert_level_to_tag(:fatal), do: 1

  @doc "TLS wire value."
  @spec alert_level_wire_value(alert_level()) :: non_neg_integer()
  def alert_level_wire_value(:warning), do: 1
  def alert_level_wire_value(:fatal), do: 2

  # ===========================================================================
  # AlertDescription (RFC 8446 Section 6.2, provisional tags 0-13)
  # ===========================================================================

  @typedoc "TLS alert descriptions (RFC 8446 Section 6.2, subset)."
  @type alert_description ::
          :close_notify | :unexpected_message | :bad_record_mac | :record_overflow
          | :handshake_failure | :bad_certificate | :unsupported_certificate
          | :certificate_revoked | :certificate_expired | :certificate_unknown
          | :illegal_parameter | :decode_error | :decrypt_error | :protocol_version

  @alert_desc_tags %{
    close_notify: 0, unexpected_message: 1, bad_record_mac: 2, record_overflow: 3,
    handshake_failure: 4, bad_certificate: 5, unsupported_certificate: 6,
    certificate_revoked: 7, certificate_expired: 8, certificate_unknown: 9,
    illegal_parameter: 10, decode_error: 11, decrypt_error: 12, protocol_version: 13
  }
  @tag_to_alert_desc Map.new(@alert_desc_tags, fn {k, v} -> {v, k} end)

  @alert_desc_strings %{
    close_notify: "close_notify", unexpected_message: "unexpected_message",
    bad_record_mac: "bad_record_mac", record_overflow: "record_overflow",
    handshake_failure: "handshake_failure", bad_certificate: "bad_certificate",
    unsupported_certificate: "unsupported_certificate",
    certificate_revoked: "certificate_revoked", certificate_expired: "certificate_expired",
    certificate_unknown: "certificate_unknown", illegal_parameter: "illegal_parameter",
    decode_error: "decode_error", decrypt_error: "decrypt_error",
    protocol_version: "protocol_version"
  }

  @doc "Decode from a C-ABI tag value."
  @spec alert_description_from_tag(non_neg_integer()) :: {:ok, alert_description()} | :error
  def alert_description_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 13 do
    {:ok, Map.fetch!(@tag_to_alert_desc, tag)}
  end
  def alert_description_from_tag(_tag), do: :error

  @doc "Encode to the C-ABI tag value."
  @spec alert_description_to_tag(alert_description()) :: non_neg_integer()
  def alert_description_to_tag(desc) when is_map_key(@alert_desc_tags, desc) do
    Map.fetch!(@alert_desc_tags, desc)
  end

  @doc "Human-readable alert description."
  @spec alert_description_to_string(alert_description()) :: String.t()
  def alert_description_to_string(desc) when is_map_key(@alert_desc_strings, desc) do
    Map.fetch!(@alert_desc_strings, desc)
  end

  @doc "Whether this alert is always fatal."
  @spec alert_description_fatal?(alert_description()) :: boolean()
  def alert_description_fatal?(:close_notify), do: false
  def alert_description_fatal?(_desc), do: true

  # ===========================================================================
  # HandshakeState (provisional tags 0-6)
  # ===========================================================================

  @typedoc "TLS handshake lifecycle states (TLS 1.3 full handshake)."
  @type handshake_state ::
          :start | :wait_server_hello | :wait_encrypted_extensions
          | :wait_cert_request | :wait_cert | :wait_finished | :connected

  @handshake_state_tags %{
    start: 0, wait_server_hello: 1, wait_encrypted_extensions: 2,
    wait_cert_request: 3, wait_cert: 4, wait_finished: 5, connected: 6
  }
  @tag_to_handshake_state Map.new(@handshake_state_tags, fn {k, v} -> {v, k} end)

  @doc "Decode from a C-ABI tag value."
  @spec handshake_state_from_tag(non_neg_integer()) :: {:ok, handshake_state()} | :error
  def handshake_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 6 do
    {:ok, Map.fetch!(@tag_to_handshake_state, tag)}
  end
  def handshake_state_from_tag(_tag), do: :error

  @doc "Encode to the C-ABI tag value."
  @spec handshake_state_to_tag(handshake_state()) :: non_neg_integer()
  def handshake_state_to_tag(state) when is_map_key(@handshake_state_tags, state) do
    Map.fetch!(@handshake_state_tags, state)
  end

  @doc "Whether application data can be sent in this state."
  @spec handshake_state_can_send_data?(handshake_state()) :: boolean()
  def handshake_state_can_send_data?(:connected), do: true
  def handshake_state_can_send_data?(_state), do: false

  @typedoc "Named TLS handshake transitions."
  @type handshake_transition ::
          :send_client_hello | :receive_server_hello | :receive_encrypted_extensions
          | :receive_cert_request | :receive_cert | :receive_finished
          | :skip_cert_request | :skip_cert

  # validate_handshake_transition removed: unproven reimplementation. The verified check lives in the
  # Idris2/Zig core; calling it needs FFI wiring not yet present in this binding.
  # Do not reimplement here. See docs/decisions/0003-keep-bindings-thin-abi-wrappers.md
end
